# **i. Introduction**

``Coder`` = M Naufal Indriatmoko  //  *Hactiv8 FTDS Batch 10*

Dataset used in this project is a Pokemon data, of pokemon games (not the cards or Pokemon Go). 

Acquired from several websites including major pokemon sites like pokemon.com and pokemondb. 

The dataset contains all kinds of data starting from pokemon's stats (Attack, Defense, etc) to its type(s), even the generation the pokemon was introduced into. 

The **objective** of the project is to showcase that the ``coder`` is able:
* to load data with Pandas

* to handle missing values in the dataset

* to manipulate column(s)

* to create query(s)

* to do grouping

* to do basic visualization with Pandas

# **ii. Import Library**

In [1]:
import pandas as pd
import numpy as np

# **iii. Data Loading**

In [2]:
poke = pd.read_csv('../dataset/Pokemon.csv')

In [29]:
poke.drop('#', axis=1).head(21) # Data overview

Name  Type 1  Type 2  Total  HP  Attack  Defense  \
0                   Bulbasaur   Grass  Poison    318  45      49       49   
1                     Ivysaur   Grass  Poison    405  60      62       63   
2                    Venusaur   Grass  Poison    525  80      82       83   
3       VenusaurMega Venusaur   Grass  Poison    625  80     100      123   
4                  Charmander    Fire     NaN    309  39      52       43   
5                  Charmeleon    Fire     NaN    405  58      64       58   
6                   Charizard    Fire  Flying    534  78      84       78   
7   CharizardMega Charizard X    Fire  Dragon    634  78     130      111   
8   CharizardMega Charizard Y    Fire  Flying    634  78     104       78   
9                    Squirtle   Water     NaN    314  44      48       65   
10                  Wartortle   Water     NaN    405  59      63       80   
11                  Blastoise   Water     NaN    530  79      83      100   
12    BlastoiseMega Blastoise   Water     NaN    630  79     103      120   
13                   Caterpie     Bug     NaN    195  45      30       35   
14                    Metapod     Bug     NaN    205  50      20       55   
15                 Butterfree     Bug  Flying    395  60      45       50   
16                     Weedle     Bug  Poison    195  40      35       30   
17                     Kakuna     Bug  Poison    205  45      25       50   
18                   Beedrill     Bug  Poison    395  65      90       40   
19      BeedrillMega Beedrill     Bug  Poison    495  65     150       40   
20                     Pidgey  Normal  Flying    251  40      45       40   

    Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0        65       65     45           1      False  
1        80       80     60           1      False  
2       100      100     80           1      False  
3       122      120     80           1      False  
4        60       50     65           1      False  
5        80       65     80           1      False  
6       109       85    100           1      False  
7       130       85    100           1      False  
8       159      115    100           1      False  
9        50       64     43           1      False  
10       65       80     58           1      False  
11       85      105     78           1      False  
12      135      115     78           1      False  
13       20       20     45           1      False  
14       25       25     30           1      False  
15       90       80     70           1      False  
16       20       20     50           1      False  
17       25       25     35           1      False  
18       45       80     75           1      False  
19       15       80    145           1      False  
20       35       35     56           1      False

In [5]:
poke.shape # the dataset has 800 rows and 13 columns

(800, 13)

In [35]:
poke.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


the dataset consists of 3 data types: integer, boolean, and object (string). Integer is the majority.



In [7]:
poke.describe().T  #overview of the values at each column

count       mean         std    min     25%    50%     75%    max
#           800.0  362.81375  208.343798    1.0  184.75  364.5  539.25  721.0
Total       800.0  435.10250  119.963040  180.0  330.00  450.0  515.00  780.0
HP          800.0   69.25875   25.534669    1.0   50.00   65.0   80.00  255.0
Attack      800.0   79.00125   32.457366    5.0   55.00   75.0  100.00  190.0
Defense     800.0   73.84250   31.183501    5.0   50.00   70.0   90.00  230.0
Sp. Atk     800.0   72.82000   32.722294   10.0   49.75   65.0   95.00  194.0
Sp. Def     800.0   71.90250   27.828916   20.0   50.00   70.0   90.00  230.0
Speed       800.0   68.27750   29.060474    5.0   45.00   65.0   90.00  180.0
Generation  800.0    3.32375    1.661290    1.0    2.00    3.0    5.00    6.0

In [47]:
poke.loc[:,['Name','HP', 'Attack']]

Name  HP  Attack
0                Bulbasaur  45      49
1                  Ivysaur  60      62
2                 Venusaur  80      82
3    VenusaurMega Venusaur  80     100
4               Charmander  39      52
..                     ...  ..     ...
795                Diancie  50     100
796    DiancieMega Diancie  50     160
797    HoopaHoopa Confined  80     110
798     HoopaHoopa Unbound  80     160
799              Volcanion  80     110

[800 rows x 3 columns]

In [60]:
poke[['Name', 'Type 1']]

Name   Type 1
0                Bulbasaur    Grass
1                  Ivysaur    Grass
2                 Venusaur    Grass
3    VenusaurMega Venusaur    Grass
4               Charmander     Fire
..                     ...      ...
795                Diancie     Rock
796    DiancieMega Diancie     Rock
797    HoopaHoopa Confined  Psychic
798     HoopaHoopa Unbound  Psychic
799              Volcanion     Fire

[800 rows x 2 columns]

In [62]:
poke['Type 1'].value_counts()

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

In [64]:
poke.axes

[RangeIndex(start=0, stop=800, step=1),
 Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
        'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
       dtype='object')]

In [65]:
'HP' in poke.keys()

True

In [66]:
'Bulbasaur' in poke.keys()

False

In [72]:
poke.loc[1:5]

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   
5  5             Charmeleon   Fire     NaN    405  58      64       58   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  
5       80       65     80           1      False

In [76]:
poke.drop('#',axis=1).loc[1:5]

Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
1                Ivysaur  Grass  Poison    405  60      62       63       80   
2               Venusaur  Grass  Poison    525  80      82       83      100   
3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123      122   
4             Charmander   Fire     NaN    309  39      52       43       60   
5             Charmeleon   Fire     NaN    405  58      64       58       80   

   Sp. Def  Speed  Generation  Legendary  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False  
5       65     80           1      False

query

In [79]:
tipe_fairy = poke[poke['Type 1'] == 'Fairy']

In [80]:
tipe_fairy

#        Name Type 1  Type 2  Total   HP  Attack  Defense  Sp. Atk  \
40    35    Clefairy  Fairy     NaN    323   70      45       48       60   
41    36    Clefable  Fairy     NaN    483   95      70       73       95   
187  173      Cleffa  Fairy     NaN    218   50      25       28       45   
189  175      Togepi  Fairy     NaN    245   35      20       65       40   
190  176     Togetic  Fairy  Flying    405   55      40       85       80   
225  209    Snubbull  Fairy     NaN    300   60      80       50       40   
226  210    Granbull  Fairy     NaN    450   90     120       75       60   
519  468    Togekiss  Fairy  Flying    545   85      50       95      120   
737  669     Flabébé  Fairy     NaN    303   44      38       39       61   
738  670     Floette  Fairy     NaN    371   54      45       47       75   
739  671     Florges  Fairy     NaN    552   78      65       68      112   
752  682    Spritzee  Fairy     NaN    341   78      52       60       63   
753  683  Aromatisse  Fairy     NaN    462  101      72       72       99   
754  684     Swirlix  Fairy     NaN    341   62      48       66       59   
755  685    Slurpuff  Fairy     NaN    480   82      80       86       85   
770  700     Sylveon  Fairy     NaN    525   95      65       65      110   
792  716     Xerneas  Fairy     NaN    680  126     131       95      131   

     Sp. Def  Speed  Generation  Legendary  
40        65     35           1      False  
41        90     60           1      False  
187       55     15           2      False  
189       65     20           2      False  
190      105     40           2      False  
225       40     30           2      False  
226       60     45           2      False  
519      115     80           4      False  
737       79     42           6      False  
738       98     52           6      False  
739      154     75           6      False  
752       65     23           6      False  
753       89     29           6      False  
754       57     49           6      False  
755       75     72           6      False  
770      130     60           6      False  
792       98     99           6       True

In [91]:
att170 = poke[poke['Attack'] > 170]
att170

#                     Name   Type 1    Type 2  Total   HP  Attack  \
163  150      MewtwoMega Mewtwo X  Psychic  Fighting    780  106     190   
232  214  HeracrossMega Heracross      Bug  Fighting    600   80     185   
424  383    GroudonPrimal Groudon   Ground      Fire    770  100     180   
426  384    RayquazaMega Rayquaza   Dragon    Flying    780  105     180   
429  386       DeoxysAttack Forme  Psychic       NaN    600   50     180   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
163      100      154      100    130           1       True  
232      115       40      105     75           2      False  
424      160      150       90     90           3       True  
426      100      180      100    115           3       True  
429       20      180       20    150           3       True

In [92]:
att170.shape

(5, 13)

In [96]:
chu = poke[poke['Name'].str.endswith('chu')]
chu

#     Name    Type 1 Type 2  Total  HP  Attack  Defense  Sp. Atk  \
30    25  Pikachu  Electric    NaN    320  35      55       40       50   
31    26   Raichu  Electric    NaN    485  60      90       55       90   
186  172    Pichu  Electric    NaN    205  20      40       15       35   

     Sp. Def  Speed  Generation  Legendary  
30        50     90           1      False  
31        80    110           1      False  
186       35     60           2      False

In [99]:
legend = poke[poke['Legendary']==True]
legend

#                 Name    Type 1    Type 2  Total   HP  Attack  \
156  144             Articuno       Ice    Flying    580   90      85   
157  145               Zapdos  Electric    Flying    580   90      90   
158  146              Moltres      Fire    Flying    580   90     100   
162  150               Mewtwo   Psychic       NaN    680  106     110   
163  150  MewtwoMega Mewtwo X   Psychic  Fighting    780  106     190   
..   ...                  ...       ...       ...    ...  ...     ...   
795  719              Diancie      Rock     Fairy    600   50     100   
796  719  DiancieMega Diancie      Rock     Fairy    700   50     160   
797  720  HoopaHoopa Confined   Psychic     Ghost    600   80     110   
798  720   HoopaHoopa Unbound   Psychic      Dark    680   80     160   
799  721            Volcanion      Fire     Water    600   80     110   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
156      100       95      125     85           1       True  
157       85      125       90    100           1       True  
158       90      125       85     90           1       True  
162       90      154       90    130           1       True  
163      100      154      100    130           1       True  
..       ...      ...      ...    ...         ...        ...  
795      150      100      150     50           6       True  
796      110      160      110    110           6       True  
797       60      150      130     70           6       True  
798       60      170      130     80           6       True  
799      120      130       90     70           6       True  

[65 rows x 13 columns]

In [101]:
legend_1 = legend[legend['Generation']==1]
legend_1  # legendary pokemon introduced in generation 1

#                 Name    Type 1    Type 2  Total   HP  Attack  \
156  144             Articuno       Ice    Flying    580   90      85   
157  145               Zapdos  Electric    Flying    580   90      90   
158  146              Moltres      Fire    Flying    580   90     100   
162  150               Mewtwo   Psychic       NaN    680  106     110   
163  150  MewtwoMega Mewtwo X   Psychic  Fighting    780  106     190   
164  150  MewtwoMega Mewtwo Y   Psychic       NaN    780  106     150   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
156      100       95      125     85           1       True  
157       85      125       90    100           1       True  
158       90      125       85     90           1       True  
162       90      154       90    130           1       True  
163      100      154      100    130           1       True  
164       70      194      120    140           1       True

In [102]:
legend_1_alt = poke[
                    (poke["Legendary"] == True) &
                    (poke['Generation'] == 1)
]
legend_1_alt

#                 Name    Type 1    Type 2  Total   HP  Attack  \
156  144             Articuno       Ice    Flying    580   90      85   
157  145               Zapdos  Electric    Flying    580   90      90   
158  146              Moltres      Fire    Flying    580   90     100   
162  150               Mewtwo   Psychic       NaN    680  106     110   
163  150  MewtwoMega Mewtwo X   Psychic  Fighting    780  106     190   
164  150  MewtwoMega Mewtwo Y   Psychic       NaN    780  106     150   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
156      100       95      125     85           1       True  
157       85      125       90    100           1       True  
158       90      125       85     90           1       True  
162       90      154       90    130           1       True  
163      100      154      100    130           1       True  
164       70      194      120    140           1       True

In [111]:
leg_fly_1 = poke [
                    (poke['Legendary']==True) &
                    (poke['Generation'] == 1) &
                    ((poke['Type 1'] == 'Flying') |
                    (poke['Type 2'] == 'Flying'))
]
leg_fly_1  # legend flying pokemon in Gen 1

#      Name    Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
156  144  Articuno       Ice  Flying    580  90      85      100       95   
157  145    Zapdos  Electric  Flying    580  90      90       85      125   
158  146   Moltres      Fire  Flying    580  90     100       90      125   

     Sp. Def  Speed  Generation  Legendary  
156      125     85           1       True  
157       90    100           1       True  
158       85     90           1       True

In [19]:
LF2 = poke [
                    (poke['Legendary']==True) &
                    (poke['Generation'] == 5) &
                    ((poke['Type_1'] == 'Flying') |
                    (poke['Type_2'] == 'Flying'))
]
LF2  # legend flying pokemon in Gen 5

#                      Name    Type_1  Type_2  Total  HP  Attack  \
702  641   TornadusIncarnate Forme    Flying     NaN    580  79     115   
703  641     TornadusTherian Forme    Flying     NaN    580  79     100   
704  642  ThundurusIncarnate Forme  Electric  Flying    580  79     115   
705  642    ThundurusTherian Forme  Electric  Flying    580  79     105   
708  645   LandorusIncarnate Forme    Ground  Flying    600  89     125   
709  645     LandorusTherian Forme    Ground  Flying    600  89     145   

     Defense  Sp_Atk  Sp_Def  Speed  Generation  Legendary  
702       70     125      80    111           5       True  
703       80     110      90    121           5       True  
704       70     125      80    111           5       True  
705       70     145      80    101           5       True  
708       90     115      80    101           5       True  
709       90     105      80     91           5       True

In [123]:
poke.query('Legendary == True &\
            Generation == 1')

#                 Name    Type 1    Type 2  Total   HP  Attack  \
156  144             Articuno       Ice    Flying    580   90      85   
157  145               Zapdos  Electric    Flying    580   90      90   
158  146              Moltres      Fire    Flying    580   90     100   
162  150               Mewtwo   Psychic       NaN    680  106     110   
163  150  MewtwoMega Mewtwo X   Psychic  Fighting    780  106     190   
164  150  MewtwoMega Mewtwo Y   Psychic       NaN    780  106     150   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
156      100       95      125     85           1       True  
157       85      125       90    100           1       True  
158       90      125       85     90           1       True  
162       90      154       90    130           1       True  
163      100      154      100    130           1       True  
164       70      194      120    140           1       True

In [130]:
poke.query('Legendary == True & Generation == 1')

#                 Name    Type 1    Type 2  Total   HP  Attack  \
156  144             Articuno       Ice    Flying    580   90      85   
157  145               Zapdos  Electric    Flying    580   90      90   
158  146              Moltres      Fire    Flying    580   90     100   
162  150               Mewtwo   Psychic       NaN    680  106     110   
163  150  MewtwoMega Mewtwo X   Psychic  Fighting    780  106     190   
164  150  MewtwoMega Mewtwo Y   Psychic       NaN    780  106     150   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
156      100       95      125     85           1       True  
157       85      125       90    100           1       True  
158       90      125       85     90           1       True  
162       90      154       90    130           1       True  
163      100      154      100    130           1       True  
164       70      194      120    140           1       True

In [4]:
poke.query('Legendary == True &\
            Generation == 1 &\
            HP > 100')

#                 Name   Type 1    Type 2  Total   HP  Attack  Defense  \
162  150               Mewtwo  Psychic       NaN    680  106     110       90   
163  150  MewtwoMega Mewtwo X  Psychic  Fighting    780  106     190      100   
164  150  MewtwoMega Mewtwo Y  Psychic       NaN    780  106     150       70   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
162      154       90    130           1       True  
163      154      100    130           1       True  
164      194      120    140           1       True

In [5]:
poke.rename(columns={'Type 1': 'Type_1', 'Type 2': 'Type_2', 'Sp. Atk': 'Sp_Atk', 'Sp. Def': 'Sp_Def'}, inplace=True)

In [13]:
poke

#                   Name   Type_1  Type_2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp_Atk  Sp_Def  Speed  Generation  Legendary  
0        65      65     45           1      False  
1        80      80     60           1      False  
2       100     100     80           1      False  
3       122     120     80           1      False  
4        60      50     65           1      False  
..      ...     ...    ...         ...        ...  
795     100     150     50           6       True  
796     160     110    110           6       True  
797     150     130     70           6       True  
798     170     130     80           6       True  
799     130      90     70           6       True  

[800 rows x 13 columns]

In [22]:
poke.query('Legendary == True &\
            Generation == 5 &\
            (Type_1 == "Flying" | Type_2 == "Flying")')

#                      Name    Type_1  Type_2  Total  HP  Attack  \
702  641   TornadusIncarnate Forme    Flying     NaN    580  79     115   
703  641     TornadusTherian Forme    Flying     NaN    580  79     100   
704  642  ThundurusIncarnate Forme  Electric  Flying    580  79     115   
705  642    ThundurusTherian Forme  Electric  Flying    580  79     105   
708  645   LandorusIncarnate Forme    Ground  Flying    600  89     125   
709  645     LandorusTherian Forme    Ground  Flying    600  89     145   

     Defense  Sp_Atk  Sp_Def  Speed  Generation  Legendary  
702       70     125      80    111           5       True  
703       80     110      90    121           5       True  
704       70     125      80    111           5       True  
705       70     145      80    101           5       True  
708       90     115      80    101           5       True  
709       90     105      80     91           5       True

In [36]:
poke['Type_2'].isnull().sum()

386

#### Grouping

In [67]:
poke.tail()

#                 Name   Type_1 Type_2  Total  HP  Attack  Defense  \
795  719              Diancie     Rock  Fairy    600  50     100      150   
796  719  DiancieMega Diancie     Rock  Fairy    700  50     160      110   
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   
799  721            Volcanion     Fire  Water    600  80     110      120   

     Sp_Atk  Sp_Def  Speed  Generation  Legendary  
795     100     150     50           6       True  
796     160     110    110           6       True  
797     150     130     70           6       True  
798     170     130     80           6       True  
799     130      90     70           6       True

In [42]:
poke.groupby('Type_1')['Speed'].mean()

Type_1
Bug          61.681159
Dark         76.161290
Dragon       83.031250
Electric     84.500000
Fairy        48.588235
Fighting     66.074074
Fire         74.442308
Flying      102.500000
Ghost        64.343750
Grass        61.928571
Ground       63.906250
Ice          63.458333
Normal       71.551020
Poison       63.571429
Psychic      81.491228
Rock         55.909091
Steel        55.259259
Water        65.964286
Name: Speed, dtype: float64

In [46]:
poke.groupby('Type_2')['Speed'].mean()

Type_2
Bug         61.666667
Dark        75.350000
Dragon      80.555556
Electric    71.500000
Fairy       61.217391
Fighting    87.384615
Fire        79.666667
Flying      85.721649
Ghost       53.857143
Grass       57.320000
Ground      54.657143
Ice         74.642857
Normal      89.250000
Poison      66.647059
Psychic     69.484848
Rock        36.000000
Steel       59.363636
Water       57.142857
Name: Speed, dtype: float64

In [83]:
poke[
        (poke['Type_1'] == 'Normal') &
        (poke['Speed'] >= 100)
].groupby(['Generation', 'Speed'])['Name'].count()
# Pokemon di tiap Generation dgn Type_1 Normal dan Speed >= 100

Generation  Speed
1           100      3
            101      1
            110      1
            115      1
            121      1
2           100      1
3           100      2
            125      1
4           100      2
            105      1
            112      1
            115      1
            120      1
            135      1
5           115      1
            128      1
6           102      1
Name: Name, dtype: int64

In [84]:
poke[
        ((poke['Type_1'] == 'Fire') |
        (poke['Type_2'] == 'Fire')) &
        (poke['Speed'] >= 100)
].groupby(['Legendary', 'Speed'])['Name'].count()
# Pokemon Legendary yg punya atribut Fire dengan Speed >= 100

Legendary  Speed
False      100      7
           101      1
           104      1
           105      1
           106      1
           108      1
           115      1
           126      1
True       100      2
Name: Name, dtype: int64

In [92]:
poke[
        ((poke['Type_1'] == 'Fire') |
        (poke['Type_2'] == 'Fire')) &
        (poke['Speed'] >= 100)
].groupby(['Generation', 'Speed'])['Name'].count()
# Pokemon yg punya atribut Fire dengan Speed >= 100 di tiap generasi

Generation  Speed
1           100      4
            105      1
2           100      2
            115      1
3           100      1
4           108      1
5           100      2
            101      1
6           104      1
            106      1
            126      1
Name: Name, dtype: int64

In [108]:
poke[
        (poke['Total'] >= 600)
].groupby(['Generation', 'Name'])['#'].count()
# Pokemon terkuat Total >= 600 di tiap generasi

Generation  Name                     
1           AerodactylMega Aerodactyl    1
            BlastoiseMega Blastoise      1
            CharizardMega Charizard X    1
            CharizardMega Charizard Y    1
            Dragonite                    1
                                        ..
6           HoopaHoopa Unbound           1
            Volcanion                    1
            Xerneas                      1
            Yveltal                      1
            Zygarde50% Forme             1
Name: #, Length: 85, dtype: int64

### Missing Values

In [93]:
poke.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type_1      800 non-null    object
 3   Type_2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp_Atk      800 non-null    int64 
 9   Sp_Def      800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


Missing values only appear in column Type_2.

There are about 800 - 414 = 386 missing values in this column.

If we ignore the missing values:

In [97]:
pk_no_miss = poke.dropna()
pk_no_miss.shape

(414, 13)

now the dataset only have 414 rows, instead of 800

Or if we ignore problematic column for example Type_2:

In [99]:
pk_no_T2 = poke.dropna(axis=1)
pk_no_T2.shape

(800, 12)

In [100]:
pk_no_T2

#                   Name   Type_1  Total  HP  Attack  Defense  Sp_Atk  \
0      1              Bulbasaur    Grass    318  45      49       49      65   
1      2                Ivysaur    Grass    405  60      62       63      80   
2      3               Venusaur    Grass    525  80      82       83     100   
3      3  VenusaurMega Venusaur    Grass    625  80     100      123     122   
4      4             Charmander     Fire    309  39      52       43      60   
..   ...                    ...      ...    ...  ..     ...      ...     ...   
795  719                Diancie     Rock    600  50     100      150     100   
796  719    DiancieMega Diancie     Rock    700  50     160      110     160   
797  720    HoopaHoopa Confined  Psychic    600  80     110       60     150   
798  720     HoopaHoopa Unbound  Psychic    680  80     160       60     170   
799  721              Volcanion     Fire    600  80     110      120     130   

     Sp_Def  Speed  Generation  Legendary  
0        65     45           1      False  
1        80     60           1      False  
2       100     80           1      False  
3       120     80           1      False  
4        50     65           1      False  
..      ...    ...         ...        ...  
795     150     50           6       True  
796     110    110           6       True  
797     130     70           6       True  
798     130     80           6       True  
799      90     70           6       True  

[800 rows x 12 columns]

# **iv. Data Cleaning**

In [8]:
a = poke.isnull()
print(a)

         #   Name  Type 1  Type 2  Total     HP  Attack  Defense  Sp. Atk  \
0    False  False   False   False  False  False   False    False    False   
1    False  False   False   False  False  False   False    False    False   
2    False  False   False   False  False  False   False    False    False   
3    False  False   False   False  False  False   False    False    False   
4    False  False   False    True  False  False   False    False    False   
..     ...    ...     ...     ...    ...    ...     ...      ...      ...   
795  False  False   False   False  False  False   False    False    False   
796  False  False   False   False  False  False   False    False    False   
797  False  False   False   False  False  False   False    False    False   
798  False  False   False   False  False  False   False    False    False   
799  False  False   False   False  False  False   False    False    False   

     Sp. Def  Speed  Generation  Legendary  
0      False  False       Fals

In [23]:
poke.loc[:,'Speed']

0       45
1       60
2       80
3       80
4       65
      ... 
795     50
796    110
797     70
798     80
799     70
Name: Speed, Length: 800, dtype: int64

# **v. Data Exploration**

# **vi. Conclusion**